wdl-lint/src/rules/input_not_sorted.rs

```diff
@@ -0,0 +1,101 @@
+//! A lint rule for sorting of inputs.
+
+use wdl_ast::span_of;
+use wdl_ast::v1::InputSection;
+use wdl_ast::AstNode;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Span;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the input not sorted rule.
+const ID: &str = "InputSorting";
+
+/// Creates a "input not sorted" diagnostic.
+fn input_not_sorted(span: Span, sorted_inputs: String) -> Diagnostic {
+    Diagnostic::warning("input not sorted")
+        .with_rule(ID)
+        .with_label("input section must be sorted".to_string(), span)
+        .with_fix(format!("sort input statements as: \n{}", sorted_inputs))
+}
+
+/// Detects unsorted input declarations.
+#[derive(Debug, Clone, Copy)]
+pub struct InputNotSortedRule;
+
+impl Rule for InputNotSortedRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that input declarations are sorted."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Each input declaration section should be sorted. Enforces an opinionated sorting. First \
+         sorts by 1. required inputs, 2. optional inputs without defaults, 3. optional inputs with \
+         defaults, and 4. non-optional inputs with a default value. Then by the type: 1. File, 2. \
+         Array[*]+, 3. Array[*], 4. struct, 5. Object, 6. Map[*, *], 7. Pair[*, *], 8. String, 9. \
+         Boolean, 10. Float, 11. Int. For ordering of the same compound type (Array[*], Map[*, *], \
+         Pair[*, *]), drop the outermost type (Array, Map, etc.) and recursively apply above \
+         sorting on the first inner type *, with ties broken by the second inner type. Continue \
+         this pattern as far as possible. Once this ordering is satisfied, it is up to the \
+         developer for final order of inputs of the same type."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style])
```

Comment:
Should we add `Clarity`?

Reply:
Probably. The only thing listed in #11 is `sorting`, which we haven't utilized.

Reply:
IMO `Sorting` would be a good addition here and also to `ImportSort`. I imagine the same people who don't intend to follow `ImportSort` won't follow `InputSort` either, so might as well give them a shorthand?